### PR TITLE
Fix fifo materialization reset

### PR DESCRIPTION
**Pull request by schilkp**
## Problem

The "Materialize FIFOs" pass generates broken FIFOs if a reset name that is not `rst` is specified.

## Repro

Input file:
<details>

`test.x`:

```rust

proc Incr {
    input: chan<s32> in;
    output: chan<s32> out;

    init { () }

    config(input: chan<s32> in, output: chan<s32> out) {
        (input, output)
    }

    next(_: ()) {
        let (tok, v) = recv(join(), input);
        send(tok, output, v + s32:1);
    }

}


proc Top {
    input: chan<s32> in;
    output: chan<s32> out;

    init { () }

    config(input: chan<s32> in, output: chan<s32> out) {
        let (s, r) = chan<s32, u32:2>("between");
        spawn Incr(input, s);
        spawn Incr(r, output);
        (input, output)
    }

    next(_: ()) { }
}
```
</details>

Then, run:

```bash
./bazel-bin/xls/dslx/ir_convert/ir_converter_main ./test.x > test.ir
./bazel-bin/xls/tools/opt_main ./test.ir --top="__test__Top_0_next" > test.opt.ir
./bazel-bin/xls/tools/codegen_main ./test.opt.ir --delay_model=unit --pipeline_stages=1 --reset="my_fancy_reset" --multi_proc --materialize_internal_fifos
```

This generates broken verilog. All normal modules feature the correct `my_fancy_reset` port:

```verilog
module __test__Tmodule __test__Top__Incr_0_next(
  input wire clk,
  input wire my_fancy_reset,
  input wire [31:0] test__input,
  input wire test__input_vld,
  input wire test__between_rdy,
  output wire [31:0] test__between,
  output wire test__between_vld,
  output wire test__input_rdy
);
```

All normal module instantiations have the correct port and connection:

```verilog
__test__Top_0_next__1 __test__Top_0_next__1_inst0 (
  .my_fancy_reset(my_fancy_reset),
  .clk(clk)
);
```

However, materialized FIFOs feature both a `rst` and `my_fancy_reset` port, and use the `my_fancy_reset` internally:

```verilog
module fifo_for_depth_2_ty_bits_32__with_bypass_register_push(
  input wire clk,
  input wire my_fancy_reset, // !!!
  input wire rst,            // !!!
  input wire push_valid,
  input wire pop_ready,
  input wire [31:0] push_data,
  output wire push_ready,
  output wire pop_valid,
  output wire [31:0] pop_data
);

// ...

if (my_fancy_reset) begin
  head <= 2'h0;
  tail <= 2'h0;
  slots <= 2'h0;
  buf__1 <= buf__1_init;
end else begin
  head <= did_push_occur ? next_head_if_push : head;
  tail <= did_pop_occur ? next_tail_if_pop : tail;
  slots <= sel_231;
  buf__1 <= did_push_occur ? array_update_206 : buf__1;
end
```

And instantiations of these instantiated FIFOs only connect to the unused `rst` port:

```verilog
  fifo_for_depth_2_ty_bits_32__with_bypass_register_push materialized_fifo_fifo_test__between_ (
    .rst(my_fancy_reset),
    .push_data(instantiation_output_149),
    .push_valid(instantiation_output_150),
    .pop_ready(instantiation_output_157),
    .push_ready(instantiation_output_240),
    .pop_data(instantiation_output_241),
    .pop_valid(instantiation_output_242),
    .clk(clk)
  );
```

Note that this works correctly if not materializing FIFOs. The `xls_fifo_wrapper` instantiations
always feature just a `rst` port that is connected correctly:

```verilog
  xls_fifo_wrapper #(
    .Width(32'd32),
    .Depth(32'd2),
    .EnableBypass(1'd1),
    .RegisterPushOutputs(1'd1),
    .RegisterPopOutputs(1'd0)
  ) fifo_test__between (
    .clk(clk),
    .rst(my_fancy_reset),
    .push_data(instantiation_output_149),
    .push_valid(instantiation_output_150),
    .pop_ready(instantiation_output_157),
    .push_ready(instantiation_output_151),
    .pop_data(instantiation_output_155),
    .pop_valid(instantiation_output_156)
  );
```

## Fix

I assume it is considered correct that `xls_fifo_wrapper` always features a `rst` port, since this is an external verilog dependency.

This patch set handles the renaming/remapping of the reset port of the instantiation (which is always named `FifoInstantiation::kResetPort`), to the reset name specified in the codegen options during fifo materialization. This ensure the materialized FIFO block behaves like any other codegen'd block and does not require any special handling.

Also adds tests that exercise this edge case.

Note that this required improvement to  `ReplaceInstantiationWith` to handle replacements with sets of ports that have different names but otherwise match in type.

I have kept this as two commits in this one PR for now. If you want me to split it up and PR it separately, please let me know.

### CODE REVIEW

File: xls/ir/block_test.cc [hunk 1]
```diff
@@ -1002,5 +1002,99 @@ block my_block(in0: bits[32], in1: bits[32], out0: bits[32], out1: bits[32]) {
...
+      block->ReplaceInstantiationWith(instantiation, inst_add, {{"a", "b"}}),
+      absl_testing::StatusIs(absl::StatusCode::kInternal,
+                             testing::ContainsRegex("name already exists")));
+
+  // Replacing with acceptable rename:
+  EXPECT_THAT(
+      block->ReplaceInstantiationWith(instantiation, inst_add,
+                                      {{"a", "a_renamed"}, {"x", "x_renamed"}}),
+      absl_testing::IsOk());
+  EXPECT_EQ(p->DumpIr(), R"(package ReplaceInstantiationWithRename
```
Comment (allight) [inline xls/ir/block_test.cc:1068]: Use golden-testfiles please.

See https://github.com/google/xls/blob/76f4486dbcaaa16705ca4fd2bd526aa24f2f45a9/xls/ir/node_util_test.cc#L295C1-L296C1 for example. This makes it easier to update if we change the way we format ir or something.

File: xls/codegen/materialize_fifos_pass.cc [hunk 2]
```diff
@@ -292,12 +293,32 @@ absl::StatusOr<bool> MaterializeFifosPass::RunInternal(
   // Intermediate list new blocks created.
 
   absl::flat_hash_map<xls::Instantiation*, Block*> impls;
+  XLS_RET_CHECK(options.codegen_options.ResetBehavior().has_value())
+      << "Reset behavior must be set to materialize fifos";
+  XLS_RET_CHECK(options.codegen_options.reset().has_value())
+      << "Fifo materialization requires reset";
+  XLS_RET_CHECK(options.codegen_options.reset().value().has_name())
+      << "Fifo materialization requires reset name";
+  auto reset_name = options.codegen_options.reset().value().name();
```
Comment (allight) [inline xls/codegen/materialize_fifos_pass.cc:302]: nit: std::string_view

**Review by allight — CHANGES_REQUESTED** [review]
Great catch. 'rst' is such a standard that this completely slipped under the radar.

**Comment by schilkp** [comment]
Thanks for the review - all addressed. 

> Use golden-testfiles please.
>
> See https://github.com/google/xls/blob/76f4486dbcaaa16705ca4fd2bd526aa24f2f45a9/xls/ir/node_util_test.cc#L295C1-L296C1 for example. This makes it easier to update if we change the way we format ir or something.

Sure - I had it like that to stay consistent with all the other tests in `block_test`.

I have added a separate commit that also refactors the rest of of the tests in `block_test` to use golden files for results to maintain sanity :)
